### PR TITLE
[kernel] E: Alloc usable frames conservatively

### PR DESCRIPTION
**Pull request by rafael0121**
## Description
This PR implements new functions to manipulate frame allocation and allow allocates usable frames conservatively, i.e., just memory marked as usable will be, all the rest, including untracked memorys are reserved.

### CODE REVIEW

File: kernel/src/mm/phys/frame.rs [hunk 1]
```diff
@@ -176,4 +208,49 @@ impl FrameAllocator {
...
+        for index in start_frame_number..=end_frame_number {
+            match self.bitmap.set(index) {
+                Ok(()) => continue,
+                Err(e) if e.code == ErrorCode::ResourceBusy => continue,
+                Err(e) => {
+                        warn!("fail to try book frame {}", index);
+                        return Err(e);
+                }
+            }
+        }
```
Comment (ppenna) [inline kernel/src/mm/phys/frame.rs:243]: If we fail in the middle and return `Err()` we would leave the system in an inconsistent state (some bits were set).

To fix this we should either:
1) Rollback what was allocated.
2) Before allocating, checking if resources are available.

File: kernel/src/mm/phys/frame.rs [hunk 2]
```diff
@@ -176,4 +208,49 @@ impl FrameAllocator {
...
+
+        Ok(())
+    }
+
+    ///
+    /// # Description
+    ///
+    /// Allocs all frames at once, without busy checks.
```
Comment (ppenna) [inline kernel/src/mm/phys/frame.rs:251]: There is a typo here. You may want to write "Allocates"

File: kernel/src/mm/phys/frame.rs [hunk 3]
```diff
@@ -176,4 +208,49 @@ impl FrameAllocator {
...
+    ///
+    pub fn fill_storage(&mut self) -> Result<(), Error> {
```
Comment (ppenna) [inline kernel/src/mm/phys/frame.rs:253]: Could you add comments regarding the return value of this function?

File: kernel/src/mm/phys/mod.rs [hunk 4]
```diff
@@ -64,6 +65,37 @@ static mut FRAME_ALLOCATOR_STORAGE: [u8; config::kernel::MEMORY_SIZE
 // Standalone Functions
 //==================================================================================================
 
+fn parse_physical_memory_regions(
```
Comment (ppenna) [inline kernel/src/mm/phys/mod.rs:68]: Could you add documentation to this function? to this function?

File: kernel/src/mm/phys/mod.rs [hunk 5]
```diff
@@ -102,6 +134,8 @@ fn book_mmio_regions(
                 Ok(()) => {},
                 // Frame lies outside addressable physical memory.
                 Err(e) if e.code == ErrorCode::InvalidArgument => {},
+                // Frame already allocated, continue try booking.
+                Err(e) if e.code == ErrorCode::ResourceBusy => {},
```
Comment (ppenna) [inline kernel/src/mm/phys/mod.rs:138]: If this happens is actually an error. It means that another region has booked it and we are just trying to supress the error.

File: kernel/src/mm/phys/mod.rs [hunk 6]
```diff
@@ -64,6 +65,37 @@ static mut FRAME_ALLOCATOR_STORAGE: [u8; config::kernel::MEMORY_SIZE
...
+    for region in physical_memory_regions {
+        match region.typ() {
+            MemoryRegionType::Usable => {
+                usable_physical_memory_regions.push_back(region);
+            },
+            MemoryRegionType::Reserved => {
+                reserved_physical_memory_regions.push_back(region);
+            }
+            _ => {
+                return Err(Error::new(ErrorCode::InvalidArgument, "Invalid memory type"));
```
Comment (ppenna) [inline kernel/src/mm/phys/mod.rs:91]: Could you please exapnd this to follow a pattern where we log the error message? Something like:

```rust
let reason: &str = "invalid memory type"
error!("parse_physical_memory_regions(): {:?}", reason)
return Err(Error::new(ErrorCode::InvalidArgument, &reason));
```

File: kernel/src/mm/phys/mod.rs [hunk 7]
```diff
@@ -73,7 +105,7 @@ fn book_physical_memory_regions(
     // Book physical memory that is not usable.
     for region in physical_memory_regions.iter() {
         info!("booking: {:?}", region);
-        frame_allocator.alloc_range(region)?;
+        frame_allocator.try_alloc_range(region)?;
```
Comment (ppenna) [inline kernel/src/mm/phys/mod.rs:108]: Why should we try to alloc isntead of allocating it?

File: kernel/src/mm/mod.rs [hunk 8]
```diff
@@ -87,19 +87,27 @@ fn parse_memory_regions(
...
                 other_virtual_memory_regions
                     .push_back(TruncatedMemoryRegion::from_memory_region(region)?);
             }
+        } else {
+            if region.typ() == MemoryRegionType::Usable {
+                if PhysicalAddress::from_virtual_address(region.start()).is_ok() {
+                    if let Ok(region) =
+                        TruncatedMemoryRegion::from_virtual_memory_region(region.clone())
+                    {
+                        physical_memory_regions.push_back(region);
```
Comment (ppenna) [inline kernel/src/mm/mod.rs:107]: We should put these in a different list, to reduce potential bugs in later parsing functions.

File: kernel/src/mm/phys/frame.rs (hunk 2 above)
Comment (rafael0121) [reply kernel/src/mm/phys/frame.rs:251]: Yes.

File: kernel/src/mm/phys/mod.rs (hunk 4 above)
Comment (rafael0121) [reply kernel/src/mm/phys/mod.rs:68]: Sure.

File: kernel/src/mm/phys/mod.rs (hunk 6 above)
Comment (rafael0121) [reply kernel/src/mm/phys/mod.rs:91]: Done.

**Pull request closed without merging**